support bypass for azure git

daxu7509 opened the pull request.

shayki5 commented on task/task.json:
```diff
@@ -210,6 +210,35 @@
+    {
+      "name": "bypassPolicy",
+      "type": "boolean",
+      "label": "bypassPolicy",
```
Please do the label with capitals and spaces

shayki5 commented on task/task.json:
```diff
@@ -210,6 +210,35 @@
+    {
+      "name": "bypassReason",
+      "type": "string",
+      "label": "bypass reason",
```
Please do the label with capitals letter

shayki5 commented on task/task.json:
```diff
@@ -210,6 +210,35 @@
+    {
+      "name": "treatAsWarning",
+      "type": "boolean",
+      "label": "treatAsWarning",
```
Please do the label with capitals letter and spcaes

shayki5 commented on task/createPullRequest.ps1:
```diff
@@ -260,6 +268,10 @@ function CreateAzureDevOpsPullRequest() {
         [string]$teamProject,
         [string]$repositoryName,
         [bool]$passPullRequestIdBackToADO
```
you forgot to put `,` in `[bool]$passPullRequestIdBackToADO`

shayki5 commented on task/createPullRequest.ps1:
```diff
@@ -324,14 +336,14 @@ function CreateAzureDevOpsPullRequest() {
             # If set auto aomplete is true 
             if ($autoComplete) {
-                SetAutoComplete -teamProject $teamProject -repositoryName $repositoryName -pullRequestId $pullRequestId -buildUserId $currentUserId -mergeStrategy $mergeStrategy -deleteSourch $deleteSourch -commitMessage $commitMessage -transitionWorkItems $transitionWorkItems
+                SetAutoComplete -teamProject $teamProject -repositoryName $repositoryName -pullRequestId $pullRequestId -buildUserId $currentUserId -mergeStrategy $mergeStrategy -deleteSourch $deleteSourch -commitMessage $commitMessage -transitionWorkItems $transitionWorkItems $bypassPolicy $bypassReason $treatAsWarning
```
please update to: 
`-bypassPolicy  $bypassPolicy -bypassReason  $bypassReason -treatAsWarning $treatAsWarning`

shayki5 commented on task/createPullRequest.ps1:
```diff
@@ -114,11 +119,14 @@ function CreatePullRequest() {
 
     if ($repoType -eq "Azure DevOps") { 
-        CreateAzureDevOpsPullRequest -teamProject $teamProject -repositoryName $repositoryName -sourceBranch $sourceBranch -targetBranch $targetBranch -title $title -description $description -reviewers $reviewers -isDraft $isDraft -autoComplete $autoComplete -mergeStrategy $mergeStrategy -deleteSourch $deleteSourch -commitMessage $commitMessage -transitionWorkItems $transitionWorkItems -linkWorkI ...
+        CreateAzureDevOpsPullRequest -teamProject $teamProject -repositoryName $repositoryName -sourceBranch $sourceBranch -targetBranch $targetBranch -title $title -description $description -reviewers $reviewers -isDraft $isDraft -autoComplete $autoComplete -mergeStrategy $mergeStrategy -deleteSourch $deleteSourch -commitMessage $commitMessage -transitionWorkItems $transitionWorkItems -linkWorkI ...
```
please update to: 
`-bypassPolicy  $bypassPolicy -bypassReason  $bypassReason -treatAsWarning $treatAsWarning`

shayki5 commented on task/createPullRequest.ps1:
```diff
@@ -85,7 +90,7 @@ function RunTask {
 
         foreach($branch in $targetBranches) {
-            CreatePullRequest -teamProject $teamProject -repositoryName $repositoryName -sourceBranch $sourceBranch -targetBranch $branch -title $title -description $description -reviewers $reviewers -repoType $repoType -isDraft $isDraft -autoComplete $autoComplete -mergeStrategy $mergeStrategy -deleteSourch $deleteSourch -commitMessage $commitMessage -transitionWorkItems $transitionWorkItems -li ...
+            CreatePullRequest -teamProject $teamProject -repositoryName $repositoryName -sourceBranch $sourceBranch -targetBranch $branch -title $title -description $description -reviewers $reviewers -repoType $repoType -isDraft $isDraft -autoComplete $autoComplete -mergeStrategy $mergeStrategy -deleteSourch $deleteSourch -commitMessage $commitMessage -transitionWorkItems $transitionWorkItems -li ...
```
please update to: 
`-bypassPolicy  $bypassPolicy -bypassReason  $bypassReason -treatAsWarning $treatAsWarning`

shayki5 commented on task/task.json:
```diff
@@ -210,6 +210,35 @@
+      "label": "bypass reason",
+      "defaultValue": "",
+      "required": false,
+      "helpMarkDown": "the reson to bypass policy.",
```
change to `The reason to the bypass policy`

shayki5 reviewed (commented):
Hi, thank you for the pull request!
I posted some comments, please check.
And - you didn't handle the warning when PR already exists. you only did it in the auto-complete.
If you want, you can check line 333 and add if-else there also.

daxu7509 commented:
> Hi, thank you for the pull request!
> I posted some comments, please check.
> And - you didn't handle the warning when PR already exists. you only did it in the auto-complete.
> If you want, you can check line 333 and add if-else there also.

hi, I updated the code now.
About warning when PR already exists, I thought this is the logic
       # If the error contains TF401179 it's mean that there is alredy a PR for the branches, so I display a warning
        if ($_ -match "TF401179" -and $treatAsWarning) {
            Write-Warning $_
        }

I added  $treatAsWarning

shayki5 commented:
Yes! I missed it. it should work.
But, I prefer to put the default to "false" for not breaking current builds.

daxu7509 commented:
> Yes! I missed it. it should work.
> But, I prefer to put the default to "false" for not breaking current builds.

sure, think you can just change that

shayki5 commented:
I made some changes to your repo. 
I switch the error/warning flag only to the option that the same PR exists.

daxu7509 commented:
> I made some changes to your repo.
> I switch the error/warning flag only to the option that the same PR exists.

sure. Can you update your repo and build a new version? Wants to try it on my azure dev op.

thanks

shayki5 commented:
Sure, I just test it.
I see the bypass can't be with the auto-complete so I separated them.
I will update you when it works and it's in production.

daxu7509 commented:
> Sure, I just test it.
> I see the bypass can't be with the auto-complete so I separated them.
> I will update you when it works and it's in production.

oh, really? I put it together with auto complete as it is in GitPullRequestCompletionOptions.
https://docs.microsoft.com/en-us/rest/api/azure/devops/git/pull%20requests/create?view=azure-devops-rest-6.0

Guess I was wrong then

shayki5 commented:
I got this:
`##[warning]{"$id":"1","innerException":null,"message":"Invalid argument value.\r\nParameter name: The bypass option cannot be used with auto-complete.","typeName":"Microsoft.TeamFoundation.SourceControl.WebServer.InvalidArgumentValueException, Microsoft.TeamFoundation.SourceControl.WebServer","typeKey":"InvalidArgumentValueException","errorCode":0,"eventId":0}`
So I guess it should be in the CompletionOptions but without autocomplete. I will check it.

shayki5 commented:
We need to take out the bypass option from the `SetAutoComplete` function, I can do it but I don't have time now :(

daxu7509 commented:
> We need to take out the bypass option from the `SetAutoComplete` function, I can do it but I don't have time now :(

hi,
if you can show me how you run this powershell locally, I can debug it for you. I actually think autocomplete should be compatible with bypaas (at least didn't really any doc saying it shouldn't)

daxu7509 commented:
> We need to take out the bypass option from the `SetAutoComplete` function, I can do it but I don't have time now :(

Hi I worked it out. Apparently you can't set auto complete together with by pass.

I played with azure cli, after setting auto complete, I need to call another command az repos pr set-vote --id xxx --vote approve

This will trigger the pr to close and code merged

shayki5 commented:
@daxu7509 I updated the code and separated the bypass from the autocomplete, but it looks like is not do anything, the PR created with:

```
{
    "sourceRefName":  "refs/heads/TestSame",
    "title":  "test new",
    "WorkItemRefs":  [
                         null
                     ],
    "description":  "",
    "completionOptions":  {
                              "bypassReason":  "test",
                              "bypassPolicy":  "true"
                          },
    "reviewers":  "",
    "isDraft":  "False",
    "targetRefName":  "refs/heads/master"
}
```
But, according to the azure-cli docs: `Bypass required policies (if any) and completes the pull request once it can be merged.` I should see the pr is should be auto completed but I don't see it.

I can share with you the pre-release extension version so you can change the code and test it. what is your organization name?

daxu7509 commented:
> @daxu7509 I updated the code and separated the bypass from the autocomplete, but it looks like is not do anything, the PR created with:
> 
> ```
> {
>     "sourceRefName":  "refs/heads/TestSame",
>     "title":  "test new",
>     "WorkItemRefs":  [
>                          null
>                      ],
>     "description":  "",
>     "completionOptions":  {
>                               "bypassReason":  "test",
>                               "bypassPolicy":  "true"
>                           },
>     "reviewers":  "",
>     "isDraft":  "False",
>     "targetRefName":  "refs/heads/master"
> }
> ```
> 
> But, according to the azure-cli docs: `Bypass required policies (if any) and completes the pull request once it can be merged.` I should see the pr is should be auto completed but I don't see it.
> 
> I can share with you the pre-release extension version so you can change the code and test it. what is your organization's name?

Hi, you need to call the PR vote method. In azure cli, it is : az repos pr set-vote
This completes PR in our case.

shayki5 commented:
> > @daxu7509 I updated the code and separated the bypass from the autocomplete, but it looks like is not do anything, the PR created with:
> > ```
> > {
> >     "sourceRefName":  "refs/heads/TestSame",
> >     "title":  "test new",
> >     "WorkItemRefs":  [
> >                          null
> >                      ],
> >     "description":  "",
> >     "completionOptions":  {
> >                               "bypassReason":  "test",
> >                               "bypassPolicy":  "true"
> >                           },
> >     "reviewers":  "",
> >     "isDraft":  "False",
> >     "targetRefName":  "refs/heads/master"
> > }
> > ```
> > 
> > 
> > But, according to the azure-cli docs: `Bypass required policies (if any) and completes the pull request once it can be merged.` I should see the pr is should be auto completed but I don't see it.
> > I can share with you the pre-release extension version so you can change the code and test it. what is your organization's name?
> 
> Hi, you need to call the PR vote method. In azure cli, it is : az repos pr set-vote
> This completes PR in our case.

So after I update the PR with the vote is it completed?

shayki5 commented:
Since it doesn't work I closing the PR for now.
